pyscf/neo/tdgrad.py

```diff
@@ -0,0 +1,979 @@
+
+        de[k] += numpy.einsum('xpq,pq->x', h1ao_e, as_dm1)
+        de[k] += numpy.einsum('xij,ij->x', z1ao_e, dm_e[p0:p1])
+
+        de[k] += td_grad.extra_force(ka, locals())
+
+    log.timer('CNEO-TDUKS nuclear gradients', *time0)
+    return de
+
+
+def as_scanner(td_grad, state=1):
+    '''
+    Modified from grad.tdrhf.as_scanner
+    '''
+    from pyscf import neo
+    if isinstance(td_grad, lib.GradScanner):
+        return td_grad
+    
+    if state == 0:
+        return td_grad.base._scf.nuc_grad_method().as_scanner()
+
+    logger.info(td_grad, 'Create scanner for %s', td_grad.__class__)
+
+    class CTDSCF_GradScanner(td_grad.__class__, lib.GradScanner):
+        def __init__(self, g, state):
+            lib.GradScanner.__init__(self, g)
+            if state is not None:
+                self.state = state
+            self._keys = self._keys.union(['e_tot'])
+        def __call__(self, mol_or_geom, state=None, **kwargs):
+            if isinstance(mol_or_geom, neo.Mole):
+                mol = mol_or_geom
+            else:
+                mol = self.mol.set_geom_(mol_or_geom, inplace=False)
+            self.reset(mol)
+
+            if state is None:
+                state = self.state
+            else:
+                self.state = state
+
+            td_scanner = self.base
+            td_scanner(mol)
+            de = self.kernel(state=state, **kwargs)
+            e_tot = self.e_tot[state-1]
+            return e_tot, de
+        @property
+        def converged(self):
+            td_scanner = self.base
+            return all((td_scanner._scf.converged,
+                        td_scanner.converged[self.state]))
+
+    return CTDSCF_GradScanner(td_grad, state)
+    
+
+class Gradients(tdrhf.Gradients):
+    ''' Analytic gradients for frozen nuclear orbital CNEO-TDDFT
+
+    Examples::
+
+    >>> from pyscf import neo
+    >>> from pyscf.neo import ctddft, tdgrad
+    >>> mol = neo.M(atom='H 0 0 0; C 0 0 1.067; N 0 0 2.213', basis='631g', 
+                    quantum_nuc = ['H'], nuc_basis = 'pb4d')
+    >>> mf = neo.CDFT(mol, xc='hf')
+    >>> mf.scf()
+    >>> td_mf = ctddft.CTDDFT(mf)
+    >>> td_mf.kernel(nstates=5)
+    >>> td_grad = tdgrad.Gradients(td_mf)
+    >>> td_grad.kernel()
+    '''
+
+    def grad_elec(self, xy, singlet, atmlst=None):
+        '''Electronic and quantum nuclear part of CNEO-TDDFT nuclear gradients'''
+
+        mf = self.base._scf
+        if isinstance(mf.components['e'], scf.uhf.UHF):
+            return grad_elec_uhf(self, xy, atmlst, self.max_memory, self.verbose)
+        else:
+            return grad_elec_rhf(self, xy, singlet, atmlst, self.max_memory, self.verbose)
+    
+    def grad_nuc(self, mol=None, atmlst=None):
+        mf_grad = self.base._scf.components['e'].nuc_grad_method()
+        return mf_grad.grad_nuc(mol, atmlst)
+        
+    def kernel(self, xy=None, state=None, singlet=True, atmlst=None):
```

Comment:
Probably you don't need to update kernel either. The only difference I see is the symmetry part, maybe that works just fine.